### PR TITLE
split out shared testing code into new subproject

### PR DESCRIPTION
**Pull request by SethTisue**
sequel to #7817

this is needed in order to avoid using "compile->test" which
we can't replicate in the IntelliJ project (we don't think)

it's better this way anyway, the partest->junit dependency was
unnecessarily thick, this replaces it with a thin dependency. I had
strongly considered doing this anyway as part of #7817

### CODE REVIEW

**Comment by SethTisue** [comment]
before merge, we need to add in (and test) the necessary IntelliJ setup changes

**Comment by SethTisue** [comment]
whoops,  now `junit/test` tests nothing. working on it...

**Comment by SethTisue** [comment]
so far my approach is "put the minimum amount of code in `testing`", to get a first working version green quickly

but it would probably be better if I went on and moved _as much code as possible_ out of `junit`. basically anything that isn't actually a test. since any test utility code is potentially useful in partests too

**Comment by dwijnand** [comment]
👨‍🎨 You _could_ follow Lightbend's (organic) naming convention and call it "testkit" 🙂

**Comment by SethTisue** [comment]
> moved as much code as possible out of junit

hmm, some of the support code seems pretty specific to the JUnit-based tests. so I'm not going to move all of it, just the ones that seem plausibly useful elsewhere. (anyway, not a high-stakes decision, we can always shuffle this stuff around further.)

**Comment by SethTisue** [comment]
> You could follow Lightbend's (organic) naming convention and call it "testkit"

hmm. the package name is `scala.tools.testing`. I do like `testkit`, but we'd want to rename the package too, I think.

**Comment by hrhino** [comment]
a tistkit, a testkit...

File: build.sbt [hunk 1]
```diff
@@ -614,8 +614,25 @@ lazy val bench = project.in(file("test") / "benchmarks")
   ).settings(inConfig(JmhPlugin.JmhKeys.Jmh)(scalabuild.JitWatchFilePlugin.jitwatchSettings))
 
 
+// at present (March 2019) there is very little code in here, because it's new.
+// in addition to providing a place for shared code, this also lets the JUnit
+// dependency flow into the partest subproject, allowing partests to (for example)
+// use the JUnit assert APIs (an also as yet little-used capability, because also new)
```
Comment (lrytz) [inline build.sbt]: we could probably live without that comment :-)

**Review by lrytz — APPROVED** [review]
squash and go

File: build.sbt (hunk 1 above)
Comment (SethTisue) [reply build.sbt]: removed

**Comment by SethTisue** [comment]
squashed — yes, let's merge on the fast track, once CI likes it, to get IntelliJ working again. we can always refine further later.

**Comment by SethTisue** [comment]
@som-snytt you might want to take a quick look and make sure I didn't misunderstand `assert8`

a tistkit, that's what our kiwi contributors call it

**Comment by som-snytt** [comment]
@hrhino tsk, tsk.

**Comment by som-snytt** [comment]
I recorded my vote against the package name `testkit`. A `testkit` tests for conformances.

@SethTisue https://github.com/scala/scala/commit/834e29ff7f2047054e4af4c05311943bcbe480a7#commitcomment-4094652

**Comment by SethTisue** [comment]
`testSupport` is an okay sbt project name, but then we would still need a package name. `test-support`? meh

I don't understand what you mean by "conformances".

to me the main problem with `testing`, as opposed to e.g. `testkit`, is that's a hard to talk about out loud without rampant ambiguity.

**Comment by SethTisue** [comment]
(do you dislike the name `testkit` in Akka and a number of other Scala-based projects too... or, you think it's fine there but just not right here?)

**Comment by som-snytt** [comment]
I don't remember adding `assert8` there, but it seems it could now be accessed from:

test/files/jvm/javaReflection/Test.scala

Edit: I have a feeling retronym suggested assert8 and other fixes at that time.

**Comment by som-snytt** [comment]
I think they use "testkit" as a rig for running a suite. Isn't it a set of tests for a feature or framework? as opposed to a general utility or framework for testing.

How about "scala enhanced test housing" or `seth`.

**Comment by SethTisue** [comment]
Akka's testkit is definitely utilities-and-frameworks-for-writing-tests-with, not a particular suite. That's my understanding of the term.

Cats, same.

Let's save the name `seth` for something much more grand. (In StarLogo/NetLogo it used to mean `set heading`, until I took it out.)

**Comment by som-snytt** [comment]
OK, I'll stand aside.

https://poolforthought.com/test-kit/

**Comment by som-snytt** [comment]
Or "test kitchen", the "kitchen" for short.

**Pull request merged**